src/cdf.rs

```diff
@@ -133,18 +130,31 @@ pub struct CdfModeContext {
 }
 
 #[derive(Clone)]
-#[repr(C)]
-pub struct CdfThreadContext {
-    pub r#ref: *mut Rav1dRef,
-    pub data: CdfThreadContext_data,
-    pub progress: *mut AtomicU32,
+pub enum CdfThreadContext {
+    QCat(c_uint),
+    Cdf(Arc<CdfThreadContext_data>),
+}
+
+impl CdfThreadContext {
+    pub fn progress(&self) -> Option<&AtomicU32> {
+        match self {
+            Self::QCat(_) => None,
+            Self::Cdf(cdf) => cdf.progress.as_ref(),
+        }
+    }
+
+    pub fn cdf(&self) -> Option<&RwLock<CdfContext>> {
+        match self {
+            Self::QCat(_) => None,
+            Self::Cdf(cdf) => Some(&cdf.cdf),
+        }
+    }
 }
 
-#[derive(Clone, Copy)]
 #[repr(C)]
-pub union CdfThreadContext_data {
-    pub cdf: *mut CdfContext,
-    pub qcat: c_uint,
+pub struct CdfThreadContext_data {
+    pub cdf: RwLock<CdfContext>,
```

Comment:
Is the `RwLock` here just for interior mutability? Do we expect any contention on it? From the usage of `unwrap` when locking it I assume not?

Reply:
Yes, it is just for interior mutability, we only use `try_` variants of locking.